### PR TITLE
Various fixes

### LINKED ISSUE #845: Save Game in Skirmish triggers exception and fails to save
Save Game in Skirmish triggers exception and fails to save.

Tested on Alpine Assault (2) map as China Nuke General without enemy player.

shot_20230205_211148_1

shot_20230205_211154_2

```
 	game.dat!00494d49()	Unknown
 	game.dat![Frames below may be incorrect and/or missing, no symbols loaded for game.dat]	Unknown
>	[Inline Frame] thyme.dll!Call_Method(const unsigned int) Line 78	C++
 	thyme.dll!GameState::Xfer_Snapshot(Xfer * xfer) Line 57	C++
 	game.dat!00601c67()	Unknown
 	game.dat!00494b4b()	Unknown
 	game.dat!00492f27()	Unknown
 	game.dat!006d12fc()	Unknown
 	game.dat!004fa2e3()	Unknown
 	game.dat!004f9bb9()	Unknown
 	game.dat!004fa2e3()	Unknown
 	game.dat!005abff5()	Unknown
 	game.dat!007cfdb1()	Unknown
 	game.dat!004fa323()	Unknown
 	game.dat!004fa5bd()	Unknown
 	game.dat!0061725f()	Unknown
 	thyme.dll!MessageStream::Propagate_Messages() Line 177	C++
 	thyme.dll!GameEngine::Update() Line 178	C++
 	thyme.dll!Win32GameEngine::Update() Line 58	C++
 	thyme.dll!GameEngine::Execute() Line 548	C++
 	thyme.dll!Game_Main(int argc, char * * argv) Line 34	C++
 	thyme.dll!main(int argc, char * * argv) Line 604	C++
 	thyme.dll!Main_Func(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 625	C++
 	game.dat!008e08ac()	Unknown
 	kernel32.dll!773900f9()	Unknown
 	ntdll.dll!77c37bbe()	Unknown
 	ntdll.dll!77c37b8e()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!00650052()	Unknown
 	game.dat!00650052()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
 	game.dat!006e0069()	Unknown
```
```
Exception thrown at 0x00494D49 in game.dat: 0xC0000005: Access violation reading location 0x900004EE.
```
```
EAX = 900004C2 EBX = 00000000 ECX = 00601C67 EDX = 00494D10 ESI = 0019F7C8 EDI = 0E9E39FC EIP = 7B6E43C0 ESP = 0019F774 EBP = FFFFFFFF EFL = 00000000 
```

### LINKED ISSUE #849: Asian fonts cause a infinite loop
Discovered in PR #841

### LINKED ISSUE #853: Trees do not draw
Trees do not draw.

Load Alpine Assault (2) map. Look for dark shadows on terrain.

shot_20230210_231305_1

### PR DESCRIPTION
**Pull request by tomsons26**
Fixes #845
Fixes #849
Fixes #853

### CODE REVIEW

**Review by xezon — APPROVED** [review]

File: src/platform/w3dengine/client/w3dtreebuffer.cpp [hunk 1]
```diff
@@ -804,18 +804,18 @@ void W3DTreeBuffer::Load_Trees_In_Index_And_Vertex_Buffers(
...
 
-                                        if (color_array) {
+                                        if (color_array != nullptr) {
                                             color = color_array[j];
                                         } else {
                                             color = 0xFFFFFFFF;
                                         }
 
-                                        vertexes->diffuse = Do_Lighting(&loc, lighting, &emissive, color, 1.0f);
+                                        vertexes->diffuse = Do_Lighting(&loc, lighting, &emissive, color, 1.0);
```
Comment (jonwil) [inline src/platform/w3dengine/client/w3dtreebuffer.cpp]: This change makes no sense, 1.0f is correct here.

**Review by jonwil — APPROVED** [review]
Other than the one comment, this looks good to me.

File: src/platform/w3dengine/client/w3dtreebuffer.cpp (hunk 1 above)
Comment (xezon) [reply src/platform/w3dengine/client/w3dtreebuffer.cpp]: If I remember correctly there was a mention of Original Game using float and double as the same type, aka compiler turns both to 4 bytes?

If that is true, shouldn't all floats in game be 4 byte floats and never 8 byte doubles?

Comment (tomsons26) [reply src/platform/w3dengine/client/w3dtreebuffer.cpp]: there are doubles in the game binary, this is a case of me doing the whole function from scratch and arriving at the same code, missed the line

**Pull request merged**